### PR TITLE
add jitter to mqtt and http checkin

### LINKED ISSUE #23: Add jitter to timings to avoid connection/publish storms
There are a number of actions the agent takes in response to events that cause it to ping the platform. For a single device that's fine. But at scale, that can cause problems.

We should introduce some variance to those timings to avoid the platform being overloaded by requests.

Scenarios that need considering:

 - When a new active snapshot is set, the platform publishes to all devices. In response, all devices make an HTTP request to get the new snapshot.
 - When the device successfully connects to the broker, it publishes an initial check-in
 - ... there may be more...
 - The regular heartbeat ping (both mqtt and http) should include some variance - eg +/- 10 seconds

### PR DESCRIPTION
**Pull request by Steve-Mcl**
fixes #23 

NOTES...

A new local class `IntervalJitter` was made to keep the logic changes in the main src to a minimum. By that I mean something capable of having an initial (lower) jittered delay (for 1st checkin) and then stable jitter centered around the desired "interval" +- jitter
 
Introducing jitter was surprisingly more difficult than I first envisioned - nothing in NPM suited my needs.  The majority of examples use this **nested timeout** pattern...
```js
    start () {
        this.stopped = false
        const runner = () => {
            // Stop if this jitter was stopped.
            if (this.stopped) { return }
            const interval = addJitter(this.#min, this.#max)
            this.intervalTimer = setTimeout(() => {
                if (this.stopped) { return }
                this.#callback && this.#callback(this)
                this.#loopTimer = setTimeout(runner, 1)  // << **** nested timeout ****
            }, interval)
        }
        runner()
    }
``` 

Where every call ends up adding to the stack and over the course of time, the delay is never centered around the set "interval" delay.  In tests, most aproaches around the web and in npm end up miles off course.

The included solution uses a unique pattern to avoid this.

For example, the below code ...
```js
const j1 = new IntervalJitter()
const timings = []
let first = true
j1.on('started', () => {
    console.log('started')
})
j1.on('interval', (t) => {
    if(!first) { timings.push(t) }
    console.log(`j1 @ ${t}ms`)
    first = false 
})
j1.on('stopped', () => {
    console.log('stopped')
    console.log('Minimum delay :', Math.min(...timings))
    console.log('Maximum delay :', Math.max(...timings))
    console.log('Average delay :', timings.reduce( ( a, b ) => a + b, 0 ) / timings.length)
    console.log('Total delay   :', timings.reduce((a, b) => a + b, 0))
})
j1.start({interval: 1000, jitter: 200, firstInterval: 10, firstJitter: 100})
setTimeout(() => {  j1.stop() }, 10100)
```

results in a jitter centered firmly around the desired "interval" **without** filling the stack with nested `setTimeouts`...
```log
started
j1 @ 18ms    << first delay is between 10 ~ 100ms as required
j1 @ 1007ms
j1 @ 1119ms
j1 @ 1062ms
j1 @ 1001ms
j1 @ 965ms
j1 @ 1033ms
j1 @ 920ms
j1 @ 1058ms
j1 @ 965ms
j1 @ 887ms   << subsequent delays all hover around 1000ms as requested
stopped
Minimum delay : 887ms
Maximum delay : 1119ms
Average delay : 1001.7ms
Total delay   : 10017ms
```

And the result of `interval: 50`, `jitter: 50` over 10 secs ...

```log
Minimum delay : 3
Maximum delay : 97
Average delay : 50.43434343434343
Total delay   : 9986
```

### CODE REVIEW

**Comment by knolleary** [comment]
Can you update the readme with details of the default jitter behaviour and how it can be configured?

**Comment by Steve-Mcl** [comment]
> Can you update the readme with details of the default jitter behaviour and how it can be configured?

Hi Nick, do you mean the parent issue #23 or some other readme?

**Comment by knolleary** [comment]
I mean the repo readme: https://github.com/flowforge/flowforge-device-agent/blob/main/README.md

If the jitter adds a new config option, it needs to be documented.

**Comment by Steve-Mcl** [comment]
> Can you update the readme with details of the default jitter behaviour and how it can be configured?

Resolved in [cc403fb](https://github.com/flowforge/flowforge-device-agent/pull/35/commits/cc403fb4710a924ef477bb5636c8f76ae493879b)

**Review by knolleary — APPROVED** [review]

File: lib/IntervalJitter.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    constructor () {
+        super()
+        this.interval = 1000
+        this.jitter = 1000
+        this.firstInterval = null
+        this.firstJitter = null
+        this.callback = null
+        this.#init()
+    }
+    #stopped = true
```
Comment (knolleary) [inline lib/IntervalJitter.js:28]: If you are going to use private class members you'll need to update `ecmaVersion` to `latest` in `.eslintrc`

**Review by knolleary — CHANGES_REQUESTED** [review]

**Review by knolleary — APPROVED** [review]

**Pull request merged**